### PR TITLE
Quickstart Backbones

### PR DESCRIPTION
**Pull request by Jacobsolawetz**
# Description

This PR adds bones to the python package to support a streamlined quickstart experience:

Noteably a:

* .login()
* .load_model()
* .webcam()

## Docs

-   [ ] Docs updated? What were the changes:
- yes - in quickstart

### CODE REVIEW

**Comment by Jacobsolawetz** [comment]
```
#!pip install jupyterlab
#!pip install ipywidgets
#!jupyter labextension install @jupyter-widgets/jupyterlab-manager
#!jupyter nbextension enable --py widgetsnbextension --sys-prefix
#!jupyter nbextension enable --user --py widgetsnbextension  --sys-prefix
```

File: roboflow/config.py [hunk 1]
```diff
@@ -1,26 +1,58 @@
...
+else:
+    config = {}
+
+
+# read configuration variables conditionally
+
+
+##1. check if variable is in environment
+##2. check if variable is in config file
+##3. return default value
```
Comment (FrancescoSaverioZuppichini) [inline roboflow/config.py:22]: Please, use [python docstring in google style](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)

File: roboflow/config.py [hunk 2]
```diff
@@ -1,26 +1,58 @@
...
+def get_var_conditional(key, default):
+    if os.getenv(key) != None:
+        return os.getenv(key)
+    elif key in config.keys():
+        return config[key]
+    else:
+        return default
```
Comment (FrancescoSaverioZuppichini) [inline roboflow/config.py:29]: Please, add a comment or use a better variables name to make it easier.

For instance, what's the type of `default`?

Comment (FrancescoSaverioZuppichini) [reply roboflow/config.py:29]: for instance, I am not sure but the above code may be rewritten by

`os.environ.get(key, config.get(key, default))`

File: roboflow/config.py [hunk 3]
```diff
@@ -30,3 +62,23 @@
...
+        RF_API_KEY = None
+        for k in RF_WORKSPACES.keys():
+            workspace = RF_WORKSPACES[k]
+            if workspace["url"] == RF_WORKSPACE:
+                RF_API_KEY = workspace["apiKey"]
+    # ENV API_KEY OVERRIDE
+    if os.getenv("RF_API_KEY") != None:
+        RF_API_KEY = os.getenv("RF_API_KEY")
+
+    return RF_API_KEY
```
Comment (FrancescoSaverioZuppichini) [inline roboflow/config.py:84]: Note that general good naming practices dictate that only `GLOBAL_VARIABLES` should be named in `UPPER_CASE`, in this case this is a function that get's exectute everytime this module is imported

File: roboflow/core/version.py [hunk 4]
```diff
@@ -344,11 +365,14 @@ def deploy(self, model_type: str, model_path: str) -> None:
...
-        class_names = []
-        for i, val in enumerate(model["model"].names):
-            class_names.append((val, model["model"].names[val]))
-        class_names.sort(key=lambda x: x[0])
-        class_names = [x[1] for x in class_names]
+        if isinstance(model["model"].names, list):
+            class_names = model["model"].names
+        else:
+            class_names = []
+            for i, val in enumerate(model["model"].names):
```
Comment (FrancescoSaverioZuppichini) [inline roboflow/core/version.py:372]: `enumerate` works on an iterable, what should be the type expected here? Since we checked for `list` before

File: roboflow/models/classification.py [hunk 5]
```diff
@@ -13,7 +13,16 @@
...
+    def __init__(
+        self,
+        api_key,
+        id,
+        name=None,
+        version=None,
+        local=False,
+        colors={},
+        preprocessing={},
+    ):
```
Comment (FrancescoSaverioZuppichini) [inline roboflow/models/classification.py:25]: This is wrong, you shouldn't use mutable classes as parameters, set them to `None` and create it if is `None`

`colors=None`

Then in the `__init__`

`self.colors = {} if colors is None else colors`

File: roboflow/models/classification.py [hunk 6]
```diff
@@ -60,13 +72,19 @@ def predict(self, image_path, hosted=False):
             # Create API URL for hosted image (slightly different)
             self.api_url += "&image=" + urllib.parse.quote_plus(image_path)
             # POST to the API
-            resp = requests.get(self.api_url)
+            resp = requests.post(self.api_url)
+
+        image_dims = {"width": "0", "height": "0"}
```
Comment (FrancescoSaverioZuppichini) [inline roboflow/models/classification.py:77]: why we are sending back this information (with `zeros`)

File: roboflow/models/instance_segmentation.py [hunk 7]
```diff
@@ -3,13 +3,15 @@
 
 
 class InstanceSegmentationModel(InferenceModel):
-    def __init__(self, api_key, version_id):
+    def __init__(self, api_key, version_id, colors={}, preprocessing={}):
```
Comment (FrancescoSaverioZuppichini) [inline roboflow/models/instance_segmentation.py:6]: Same as before

File: roboflow/models/object_detection.py [hunk 8]
```diff
@@ -142,11 +150,27 @@ def predict(
...
+                        self.preprocessing["resize"]["width"]
+                    ) or dimensions[1] > int(self.preprocessing["resize"]["height"]):
+                        image = image.resize(
+                            (
+                                int(self.preprocessing["resize"]["width"]),
+                                int(self.preprocessing["resize"]["height"]),
+                            )
+                        )
+                        dimensions = image.size
+                        resize = True
```
Comment (FrancescoSaverioZuppichini) [inline roboflow/models/object_detection.py:172]: For the final version, could be either have more descriptive variables name for the condition or a little comment?

File: roboflow/models/object_detection.py [hunk 9]
```diff
@@ -181,21 +209,213 @@ def predict(
...
+        Infers detections based on webcam feed from specified model
+
+        :param webcam_id: Webcam ID (default 0)
+        :param inference_engine_url: Inference engine address to use (default https://detect.roboflow.com)
+        :param in_jupyter: Whether or not to display the webcam within Jupyter notebook (default True)
+        :param confidence: Confidence threshold for detections
+        :param overlap: Overlap threshold for detections
+        :param stroke: Stroke width for bounding box
+        :param labels: Whether to show labels on bounding box
+        """
```
Comment (FrancescoSaverioZuppichini) [inline roboflow/models/object_detection.py:288]: Why are we using `sphinx` docstring style? Is it legacy?

File: roboflow/models/object_detection.py [hunk 10]
```diff
@@ -181,21 +209,213 @@ def predict(
...
+        self.__generate_url(
+            confidence=confidence,
+            overlap=overlap,
+            stroke=stroke,
+            labels=labels,
+            inference_engine_url=inference_engine_url,
+        )
+
+        def plot_one_box(
+            x, img, color=None, label=None, line_thickness=None, colors={}
```
Comment (FrancescoSaverioZuppichini) [inline roboflow/models/object_detection.py:302]: same thing for `colors` here, as before

File: roboflow/models/object_detection.py [hunk 11]
```diff
@@ -181,21 +209,213 @@ def predict(
...
+
+            if label in colors.keys():
+                color = colors[label]
+                color = color.lstrip("#")
+                color = tuple(int(color[i : i + 2], 16) for i in (0, 2, 4))
+
+            tl = (
+                line_thickness or round(0.002 * (img.shape[0] + img.shape[1]) / 2) + 1
+            )  # line/font thickness
+            # color = color or [random.randint(0, 255) for _ in range(3)]
```
Comment (FrancescoSaverioZuppichini) [inline roboflow/models/object_detection.py:316]: leftover :)

File: roboflow/models/object_detection.py [hunk 12]
```diff
@@ -181,21 +209,213 @@ def predict(
...
+            for pred in predictions:
+                formatted_pred = [
+                    pred["x"],
+                    pred["y"],
+                    pred["x"],
+                    pred["y"],
+                    pred["confidence"],
+                ]
+
+                # convert to top-left x/y from center
```
Comment (FrancescoSaverioZuppichini) [inline roboflow/models/object_detection.py:387]: nice comment

File: roboflow/models/object_detection.py [hunk 13]
```diff
@@ -181,21 +209,213 @@ def predict(
...
+            else:
+                cv2.imshow("Roboflow Webcam Inference", frame)
+                if cv2.waitKey(1) & 0xFF == ord("q"):  # quit when 'q' is pressed
+                    cap.release()
+                    break
+
+        cap.release()
+        if not within_jupyter:
+            cv2.destroyWindow("Roboflow Webcam Inference")
+            cv2.destroyAllWindows()
```
Comment (FrancescoSaverioZuppichini) [inline roboflow/models/object_detection.py:417]: nice

File: roboflow/util/general.py [hunk 14]
```diff
@@ -0,0 +1,7 @@
+import sys
+
+
+def write_line(line):
+    sys.stdout.write("\r" + line)
+    sys.stdout.write("\n")
+    sys.stdout.flush()
```
Comment (FrancescoSaverioZuppichini) [inline roboflow/util/general.py:7]: why are we doing this?

Comment (FrancescoSaverioZuppichini) [reply roboflow/util/general.py:7]: why this file is named `general`?

File: roboflow/util/prediction.py [hunk 15]
```diff
@@ -42,7 +42,7 @@ def plot_image(image_path):
     return figure, axes
 
 
-def plot_annotation(axes, prediction=None, stroke=1, transparency=60):
+def plot_annotation(axes, prediction=None, stroke=1, transparency=60, colors={}):
```
Comment (FrancescoSaverioZuppichini) [inline roboflow/util/prediction.py:45]: same as before for `colors`

File: roboflow/util/prediction.py [hunk 16]
```diff
@@ -53,7 +53,13 @@ def plot_annotation(axes, prediction=None, stroke=1, transparency=60):
...
     # Object Detection annotation
+
+    stroke_color = "r"
+
     if prediction["prediction_type"] == OBJECT_DETECTION_MODEL:
+        if prediction["class"] in colors.keys():
+            stroke_color = colors[prediction["class"]]
+
         # Get height, width, and center coordinates of prediction
         if prediction is not None:
```
Comment (FrancescoSaverioZuppichini) [inline roboflow/util/prediction.py:64]: what will happen if `prediction` is `None`? Should it fail?

File: roboflow/util/prediction.py [hunk 17]
```diff
@@ -78,9 +84,12 @@ def plot_annotation(axes, prediction=None, stroke=1, transparency=60):
...
+        if prediction["class"] in colors.keys():
+            stroke_color = colors[prediction["class"]]
+
         points = [[p["x"], p["y"]] for p in prediction["points"]]
         polygon = patches.Polygon(
-            points, linewidth=stroke, edgecolor="r", facecolor="none"
+            points, linewidth=stroke, edgecolor=stroke_color, facecolor="none"
         )
         axes.add_patch(polygon)
     elif prediction["prediction_type"] == SEMANTIC_SEGMENTATION_MODEL:
```
Comment (FrancescoSaverioZuppichini) [inline roboflow/util/prediction.py:95]: No `else` in the end, this conditional flow seems to be incorrect

File: roboflow/util/prediction.py [hunk 18]
```diff
@@ -93,7 +102,11 @@ def plot_annotation(axes, prediction=None, stroke=1, transparency=60):
 
 class Prediction:
     def __init__(
-        self, json_prediction, image_path, prediction_type=OBJECT_DETECTION_MODEL
+        self,
+        json_prediction,
+        image_path,
+        prediction_type=OBJECT_DETECTION_MODEL,
+        colors={},
```
Comment (FrancescoSaverioZuppichini) [inline roboflow/util/prediction.py:109]: same as before

File: roboflow/util/prediction.py [hunk 19]
```diff
@@ -179,6 +197,8 @@ def save(self, output_path="predictions.jpg", stroke=2, transparency=60):
             )
 
         elif self["prediction_type"] == CLASSIFICATION_MODEL:
+            if self["top"] in self.colors.keys():
```
Comment (FrancescoSaverioZuppichini) [inline roboflow/util/prediction.py:200]: what's top?

File: roboflow/util/prediction.py [hunk 20]
```diff
@@ -473,7 +497,9 @@ def json(self):
         return prediction_group_json
 
     @staticmethod
-    def create_prediction_group(json_response, image_path, prediction_type, image_dims):
+    def create_prediction_group(
```
Comment (FrancescoSaverioZuppichini) [inline roboflow/util/prediction.py:500]: why this method is `static`? The correct way for factory methods is `classmethods`

File: roboflow/util/versions.py [hunk 21]
```diff
@@ -45,7 +45,7 @@ def print_warn_for_wrong_dependencies_versions(
     dependencies_versions: List[Tuple[str, str, str]]
 ):
     wrong_dependencies_versions = get_wrong_dependencies_versions(dependencies_versions)
-    for (dependency, order, version, module_version) in wrong_dependencies_versions:
+    for dependency, order, version, module_version in wrong_dependencies_versions:
```
Comment (FrancescoSaverioZuppichini) [inline roboflow/util/versions.py:48]: why this change?

**Review by FrancescoSaverioZuppichini — COMMENTED** [review]
I know this is a **draft**, I've added an extensive list of comments. Be sure to:

- never used immutable variables as default arguments
- when something has a big conditional flow, either name the conditional flow variable or add a comment for us

Thanks :rocket:

File: roboflow/core/version.py (hunk 4 above)
Comment (Jacobsolawetz) [reply roboflow/core/version.py:372]: In this case I think it is a dict

File: roboflow/models/classification.py (hunk 5 above)
Comment (Jacobsolawetz) [reply roboflow/models/classification.py:25]: Ok, please level me up @FrancescoSaverioZuppichini, why is that?

File: roboflow/models/classification.py (hunk 6 above)
Comment (Jacobsolawetz) [reply roboflow/models/classification.py:77]: this was probably to fix a bug - and dims usually not relevant with classification

I'll dig in

**Comment by Jacobsolawetz** [comment]
* Better plotting display
* webcam stop button
* there is a likely bug with certain class colors in COCO

File: roboflow/config.py (hunk 2 above)
Comment (Jacobsolawetz) [reply roboflow/config.py]: It may indeed be written, but I like the cascade - going to keep it with the new comments

File: roboflow/util/prediction.py (hunk 19 above)
Comment (Jacobsolawetz) [reply roboflow/util/prediction.py:202]: "top" is the predicted classification class based on our schema there

File: roboflow/util/prediction.py (hunk 20 above)
Comment (Jacobsolawetz) [reply roboflow/util/prediction.py:502]: I think this prediction.py could use a large refactor in general, I will not attempt to defend it in this PR

File: roboflow/util/versions.py (hunk 21 above)
Comment (Jacobsolawetz) [reply roboflow/util/versions.py:48]: Linter I think

File: requirements.txt [hunk 22]
```diff
@@ -18,3 +18,4 @@ tqdm>=4.41.0
 PyYAML>=5.3.1
 wget
 requests_toolbelt
+ipython>=8.9.0
```
Comment (yeldarby) [inline requirements.txt:21]: How heavy is this additional requirement? 

We probably should put some effort into really slimming down our list of requirements so installing our pip package isn’t such a burden (OpenCV and matplotlib in particular are beasts).

Comment (FrancescoSaverioZuppichini) [reply requirements.txt:21]: we can have additional installs in `setup.py`, e.g. for `notebook` is setup. Once loaded, the package can ask the runtime where he is and download additional stuff 

I think opencv is needed in general, not sure why we have `matplotlib` is we have opencv/pillow

**Comment by Jacobsolawetz** [comment]
@Jacobsolawetz test

File: roboflow/models/classification.py (hunk 6 above)
Comment (FrancescoSaverioZuppichini) [reply roboflow/models/classification.py:77]: Copy, thanks. Resolving

File: roboflow/__init__.py [hunk 23]
```diff
@@ -67,14 +69,117 @@ def auth(api_key):
...
+    if workspace is None:
+        write_line(
+            "visit " + APP_URL + "/auth-cli" " to get your authentication token."
+        )
+    else:
+        write_line(
+            "visit "
+            + APP_URL
+            + "/auth-cli/?workspace="
+            + workspace
```
Comment (yeldarby) [inline roboflow/__init__.py:93]: Brad TODO: hook into this in core repo

File: roboflow/models/classification.py (hunk 6 above)
Comment (yeldarby) [reply roboflow/models/classification.py:77]: Unresolving because it tripped me up in the notebook that this was here & had wrong values (also they're strings not numbers?)

File: roboflow/models/object_detection.py [hunk 24]
```diff
@@ -142,11 +153,28 @@ def predict(
...
                 image = Image.open(image_path).convert("RGB")
                 dimensions = image.size
+                original_dimensions = copy.deepcopy(dimensions)
+
+                # Here we resize the image to the preprocessing settings before sending it over the wire
+                if "resize" in self.preprocessing.keys():
+                    if dimensions[0] > int(
+                        self.preprocessing["resize"]["width"]
+                    ) or dimensions[1] > int(self.preprocessing["resize"]["height"]):
+                        image = image.resize(
```
Comment (yeldarby) [inline roboflow/models/object_detection.py:169]: Looks like this only does `Stretch To` but should follow the same resize strat they preproc'd with.

File: roboflow/models/object_detection.py [hunk 25]
```diff
@@ -181,21 +213,252 @@ def predict(
...
                 image_path=image_path,
                 prediction_type=OBJECT_DETECTION_MODEL,
                 image_dims=image_dims,
+                colors=self.colors,
             )
         # Returns base64 encoded Data
         elif self.format == "image":
             return resp.content
 
+    def webcam(
```
Comment (yeldarby) [inline roboflow/models/object_detection.py:270]: This is cool -- do we have regular video infer built into pip as well?

File: roboflow/__init__.py [hunk 26]
```diff
@@ -118,6 +120,8 @@ def login(workspace=None, force=False):
 
     else:
         raise RuntimeError("Error logging in")
```
Comment (FrancescoSaverioZuppichini) [inline roboflow/__init__.py:122]: This is not very description to user nor providing information if he opens a PR with the error, just call

```
r_login.raise_for_status()
```

and remove the condition

File: roboflow/__init__.py [hunk 27]
```diff
@@ -118,6 +120,8 @@ def login(workspace=None, force=False):
...
+    
+    #return Roboflow().workspace()
```
Comment (FrancescoSaverioZuppichini) [inline roboflow/__init__.py:124]: leftover

**Review by FrancescoSaverioZuppichini — CHANGES_REQUESTED** [review]
One minor comment

File: roboflow/__init__.py (hunk 23 above)
Comment (Jacobsolawetz) [reply roboflow/__init__.py:97]: @yeldarby what this? I just authed via new account in incognito and it seemed pretty good

Comment (yeldarby) [reply roboflow/__init__.py:97]: If you have multiple workspaces, selecting the one you need

Comment (Jacobsolawetz) [reply roboflow/__init__.py:97]: Ok - I think it's OK for now, new user flow will be main concern for this method

File: roboflow/models/object_detection.py (hunk 25 above)
Comment (Jacobsolawetz) [reply roboflow/models/object_detection.py:270]: We do not

File: roboflow/models/object_detection.py (hunk 24 above)
Comment (Jacobsolawetz) [reply roboflow/models/object_detection.py:169]: ^^ @paulguerrie when we get a chance let's add img_dims back to classification response in infer server, that way we can know for hosted urls - @Tyler-Odenthal fixed for local classifications

**Review by paulguerrie — APPROVED** [review]
Looks great!  Pumped about this in the quick start notebooks!
